diff -crN freeradius-server-2.1.10.org/raddb/modules/ldap freeradius-server-2.1.10/raddb/modules/ldap
*** freeradius-server-2.1.10.org/raddb/modules/ldap	2010-09-28 20:03:56.000000000 +0900
--- freeradius-server-2.1.10/raddb/modules/ldap	2011-05-07 05:33:26.000000000 +0900
***************
*** 158,163 ****
--- 158,171 ----
  	#  allowed values: {no, yes}
  	# set_auth_type = yes
  
+ 	#
+ 	# mac-radius when you only want to use.
+ 	#ldap filter to authenticate using only ignores the password check.
+ 	#The default is off. (password authentication to)
+ 	#Please enable mac-radius only when performing.
+ 	#
+ 	#set_mac_radius_mode_without_password = no
+ 
  	#  ldap_debug: debug flag for LDAP SDK
  	#  (see OpenLDAP documentation).  Set this to enable
  	#  huge amounts of LDAP debugging on the screen.
diff -crN freeradius-server-2.1.10.org/src/modules/rlm_ldap/rlm_ldap.c freeradius-server-2.1.10/src/modules/rlm_ldap/rlm_ldap.c
*** freeradius-server-2.1.10.org/src/modules/rlm_ldap/rlm_ldap.c	2010-09-28 20:03:56.000000000 +0900
--- freeradius-server-2.1.10/src/modules/rlm_ldap/rlm_ldap.c	2011-05-07 05:33:53.000000000 +0900
***************
*** 173,178 ****
--- 173,179 ----
  	int		 edir_account_policy_check;
  #endif
  	int		 set_auth_type;
+ 	int		 set_mac_radius_mode_without_password;
  
  	/*
  	 *	For keep-alives.
***************
*** 343,348 ****
--- 344,350 ----
  #endif
  
  	{"set_auth_type", PW_TYPE_BOOLEAN, offsetof(ldap_instance,set_auth_type), NULL, "yes"},
+ 	{"set_mac_radius_mode_without_password", PW_TYPE_BOOLEAN, offsetof(ldap_instance,set_mac_radius_mode_without_password), NULL, "no"},
  
  	{ "keepalive", PW_TYPE_SUBSECTION, 0, NULL, (const void *) keepalive_config },
  	{NULL, -1, 0, NULL, NULL}
***************
*** 1765,1771 ****
  	       if (!pairfind(request->config_items, PW_CLEARTEXT_PASSWORD) &&
  		   !pairfind(request->config_items, PW_USER_PASSWORD) &&
  		   !pairfind(request->config_items, PW_PASSWORD_WITH_HEADER) &&
! 		   !pairfind(request->config_items, PW_CRYPT_PASSWORD)) {
  		       DEBUG("WARNING: No \"known good\" password was found in LDAP.  Are you sure that the user is configured correctly?");
  	       }
         }
--- 1767,1774 ----
  	       if (!pairfind(request->config_items, PW_CLEARTEXT_PASSWORD) &&
  		   !pairfind(request->config_items, PW_USER_PASSWORD) &&
  		   !pairfind(request->config_items, PW_PASSWORD_WITH_HEADER) &&
! 		   !pairfind(request->config_items, PW_CRYPT_PASSWORD) &&
! 		   !inst->set_mac_radius_mode_without_password) {
  		       DEBUG("WARNING: No \"known good\" password was found in LDAP.  Are you sure that the user is configured correctly?");
  	       }
         }
***************
*** 1828,1851 ****
  		return RLM_MODULE_INVALID;
  	}
  
! 	if (!request->password){
! 		radlog(L_AUTH, "  [%s] Attribute \"User-Password\" is required for authentication.", inst->xlat_name);
! 		DEBUG2("  You seem to have set \"Auth-Type := LDAP\" somewhere.");
! 		DEBUG2("  THAT CONFIGURATION IS WRONG.  DELETE IT.");
! 		DEBUG2("  YOU ARE PREVENTING THE SERVER FROM WORKING PROPERLY.");
! 		return RLM_MODULE_INVALID;
! 	}
  
! 	if(request->password->attribute != PW_USER_PASSWORD) {
! 		radlog(L_AUTH, "  [%s] Attribute \"User-Password\" is required for authentication. Cannot use \"%s\".", inst->xlat_name, request->password->name);
! 		return RLM_MODULE_INVALID;
! 	}
  
! 	if (request->password->length == 0) {
! 		snprintf(module_fmsg,sizeof(module_fmsg),"  [%s] empty password supplied", inst->xlat_name);
! 		module_fmsg_vp = pairmake("Module-Failure-Message", module_fmsg, T_OP_EQ);
! 		pairadd(&request->packet->vps, module_fmsg_vp);
! 		return RLM_MODULE_INVALID;
  	}
  
  	/*
--- 1831,1860 ----
  		return RLM_MODULE_INVALID;
  	}
  
! 	if (inst->set_mac_radius_mode_without_password) {
! 			DEBUG("  this is set_mac_radius_mode_without_password mode.");
! 			DEBUG("  password is no check!!!!.");
! 	} else {
  
! 		if (!request->password){
! 			radlog(L_AUTH, "  [%s] Attribute \"User-Password\" is required for authentication.", inst->xlat_name);
! 			DEBUG2("  You seem to have set \"Auth-Type := LDAP\" somewhere.");
! 			DEBUG2("  THAT CONFIGURATION IS WRONG.  DELETE IT.");
! 			DEBUG2("  YOU ARE PREVENTING THE SERVER FROM WORKING PROPERLY.");
! 			return RLM_MODULE_INVALID;
! 		}
  
! 		if(request->password->attribute != PW_USER_PASSWORD) {
! 			radlog(L_AUTH, "  [%s] Attribute \"User-Password\" is required for authentication. Cannot use \"%s\".", inst->xlat_name, request->password->name);
! 			return RLM_MODULE_INVALID;
! 		}
! 	
! 		if (request->password->length == 0) {
! 			snprintf(module_fmsg,sizeof(module_fmsg),"  [%s] empty password supplied", inst->xlat_name);
! 			module_fmsg_vp = pairmake("Module-Failure-Message", module_fmsg, T_OP_EQ);
! 			pairadd(&request->packet->vps, module_fmsg_vp);
! 			return RLM_MODULE_INVALID;
! 		}
  	}
  
  	/*
***************
*** 1911,1916 ****
--- 1920,1931 ----
  
  	RDEBUG("user DN: %s", user_dn);
  
+ 	if (inst->set_mac_radius_mode_without_password) {
+ 		DEBUG("  skip password check by set_mac_radius_mode_without_password mode.");
+ 		inst->failed_conns = 0;
+ 		return RLM_MODULE_OK;
+ 	}
+ 
  #ifndef NOVELL
  	ld_user = ldap_connect(instance, user_dn, request->password->vp_strvalue,
  			       1, &res, NULL);
